### ISSUE TITLE
Support for adding subnets to existing storage account network rules

### ISSUE THREAD

**Issue by dwatts-airbus** [terraform-providers/terraform-provider-azurerm#10599]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I would like the equivalent of the "az storage account network-rule add" functionality in terraform. Today, you can create a azurerm_storage_account_network_rules resource, but it completely replaces the set of rules on the storage account. In some cases I would just like to add a subnet to an existing set of rules. Imagine a scenario where multiple developers have their own environments, but are accessing the same storage accounts. Each developer runs terraform to create their environment, and needs to add their subnet to the rules on the storage account.

I don't believe there's currently a way to handle this. When one developer runs their terraform scripts, it will replace the storage network rules, removing access from the other developer. Is it already possible to manage this in some way, or does this require a new feature in terraform?

### New or Affected Resource(s)

azurerm_storage_account_network_rules

**Comment by favoretti** [comment]
Terraform is a declarative language, it compiled the end state of the resource. If the list in the code isn't complete - how would it know what the end state should be?

I understand that you want to keep what's there and only add to it, but IMO this is contradictory to how terraform should be used.

**Comment by mouellet** [comment]
The use case defined here is valid and quite common, especially in organization using Hub & Spoke network topology. I've just opened a PR with a new resource that manages individual network rule (they are uniquely identifiable) the same way the azure cli does.

**Comment by gestrada-kr** [comment]
I agree is quite common, we have this exact problem, storage account is shared between multiple service principals and some network configuration is added by other teams. For now, we resort to have a single team handle the network rules, but this is not scalable at all.

**Comment by Amier3** [comment]
Update: We closed #14744 due to some inherent issues, and will re-evaluate with product to solve this problem. We encourage anyone coming across this to give a 👍 and add additional context in the form of a comment.

**Comment by malindhorst** [comment]
Just to warm this issue up again:

I recently came across this topic and would love to see this feature! 

In my case I need to implement a custom script extension that will be applied to a variety of Windows VMs in different subnets.  Scripts and auxiliary files should be stored centrally on a storage account managed somewhere in my organisation. We also have a Hub and Spoke / Landingzone concept. Using private endpoints is mandatory and enforced by policies. 

- I cannot allow private IP Adress ranges in the storage account access rules (that would have been some kind of workaround). 
- I also cannot allow subnets that may not exist yet as they are managed by teams within their own Landingzones throughout the company. 
 
So I am either stuck with some kind of chicken and egg problem here, or forced to duplicate the content to $n additional storage accounts, one per client Landingzone, each.
